### ISSUE TITLE
Allows zoom keyboard shortcuts (Ctrl+=, Ctrl+-) to be quietly "hijacked" by extension shortcuts (IntelliJ IDEA Keybindings)

### ISSUE THREAD

**Issue by Chealer** [microsoft/vscode#110254]
- VSCode Version: 1.51.0
- OS Version: 10

Steps to Reproduce:

1. Ensure an extension exposing the issue is enabled ([IntelliJ IDEA Keybindings](https://marketplace.visualstudio.com/items?itemName=k--kato.intellij-idea-keybindings) is one).
2. Get into a context where zoom shortcuts do not work. For example, if the Java language pack is enabled, focus a tab with a Java file editor.
3. Type the keyboard shortcut for Zoom In or Zoom Out (Ctrl + = or Ctrl + -)

Does this issue occur when all extensions are disabled?: No

**Comment by Chealer** [comment]
To complement the report, the main case I found when this happens is in a Java file, where using these shortcuts folds or unfolds code blocks (for example, if the cursor is in a function definition, _Zoom Out_ hides the implementation).

But this is more general (Java .properties files are also affected).

The proper behavior for these shortcuts is indicated in the View menu, but also [in the documentation](https://code.visualstudio.com/docs/editor/accessibility#_zoom).

**Comment by alexdima** [comment]
@Chealer Could you please try to see what commands execute when pressing your keybinding:

https://github.com/microsoft/vscode/wiki/Keybinding-Issues#basic-troubleshooting

This will help you understand what command is dispatched by our keybinding subsystem.

**Comment by Chealer** [comment]
Thank you very much @alexdima, I was not aware of this troubleshooting tool, which is very useful and allows me to effortlessly explain the problem. However, I note that I had a hard time following the instructions. I only managed to follow thanks to the screencasts you provided in ticket #37415.

The log pasted below shows my usage of the _IntelliJ IDEA Keybindings_ extension triggered the issue:

> [2020-11-16 15:18:38.337] [renderer1] [info] [KeybindingService]: / Received  keydown event - modifiers: [ctrl], code: ControlLeft, keyCode: 17, key: Control
> [2020-11-16 15:18:38.337] [renderer1] [info] [KeybindingService]: | Converted keydown event - modifiers: [ctrl], code: ControlLeft, keyCode: 5 ('Ctrl')
> [2020-11-16 15:18:38.337] [renderer1] [info] [KeybindingService]: \ Keyboard event cannot be dispatched.
> [2020-11-16 15:18:38.836] [renderer1] [info] [KeybindingService]: / Received  keydown event - modifiers: [ctrl], code: Equal, keyCode: 187, key: =
> [2020-11-16 15:18:38.836] [renderer1] [info] [KeybindingService]: | Converted keydown event - modifiers: [ctrl], code: Equal, keyCode: 81 ('=')
> [2020-11-16 15:18:38.836] [renderer1] [info] [KeybindingService]: | Resolving ctrl+=
> [2020-11-16 15:18:38.836] [renderer1] [info] [KeybindingService]: \ From 2 keybinding entries, matched editor.unfold, when: editorFocus, source: built-in extension k--kato.intellij-idea-keybindings.
> [2020-11-16 15:18:41.408] [renderer1] [info] [KeybindingService]: / Received  keydown event - modifiers: [ctrl], code: ControlLeft, keyCode: 17, key: Control
> [2020-11-16 15:18:41.408] [renderer1] [info] [KeybindingService]: | Converted keydown event - modifiers: [ctrl], code: ControlLeft, keyCode: 5 ('Ctrl')
> [2020-11-16 15:18:41.408] [renderer1] [info] [KeybindingService]: \ Keyboard event cannot be dispatched.
> [2020-11-16 15:18:41.612] [renderer1] [info] [KeybindingService]: / Received  keydown event - modifiers: [ctrl], code: Minus, keyCode: 189, key: -
> [2020-11-16 15:18:41.612] [renderer1] [info] [KeybindingService]: | Converted keydown event - modifiers: [ctrl], code: Minus, keyCode: 83 ('-')
> [2020-11-16 15:18:41.612] [renderer1] [info] [KeybindingService]: | Resolving ctrl+-
> [2020-11-16 15:18:41.612] [renderer1] [info] [KeybindingService]: \ From 2 keybinding entries, matched editor.fold, when: editorFocus, source: built-in extension k--kato.intellij-idea-keybindings.


That extension catches these shortcuts to fold and unfold code blocks, for example the contents of functions in Java files. This makes sense to imitate JetBrains, however it apparently does that broadly, even in contexts where folding is impossible, such as Java .properties files. As far as I can see, all editors are affected, no matter the file type. The original effect is only preserved in some contexts like the start page, or when browsing extensions.

So there are 2 issues here:

1. The IntelliJ IDEA Keybindings extension probably catches zoom shortcuts too broadly.
2. **VSCode does not warn about the keyboard conflicts**, either when installing the extension or when they occur.

**Please consider this ticket as a report of issue #2.** I will provide a reference here if I report issue #1 against the IntelliJ IDEA Keybindings extension.

**Comment by alexdima** [comment]
From our point of view, this is a valid extension point that is exercised by the extension. We allow extensions to override built-in keybindings **by design**. That is basically how all of the keymap extensions work.

In this case, the `IntelliJ IDEA Keybindings` makes use of that extension point and defines new keybinding rules for `ctrl+=` and `ctrl+-`, effectively shadowing our built-in keybindings. Please report this directly to the extension, as shadowing built-in keybindings is something we intentionally allow extensions to do.

**Comment by Chealer** [comment]
Thanks again @alexdima
The problematic commands are _Fold_ and _Unfold_.

I understand the mechanism and yes, I will try to report to the extension. The problem is not so much the conflicts as they being unpredictable / hard to investigate. I did mean the scare quotes around "hijacked", and although I can't format the title, the most important word is "quietly". There are several ways this could be fixed, and I don't want to mandate any one in particular:

- prevent extensions from "hijacking" these shortcuts
- warn when the default commands are shadowed, or prompt to offer rebinding
- warn when a user presses a shortcut for conflicting commands
- or stop showing the shortcut for the default command in the menu when it is shadowed, or show it differently to hint that it may not work

**Comment by alexdima** [comment]
I have considered the first 3 suggestions, but we cannot implement them. I think there is a misunderstanding of expectations here. When you install the IntelliJ IDEA keymap, you basically agree to get the keybindings as close as possible to IntelliJ IDEA. If IntelliJ IDEA does not zoom in on `ctrl+=`, then vscode configured with the IntelliJ IDEA keymap will not zoom in on `ctrl+=`. There is no reason to prevent or warn because the feature works as expected.

But that is OK. It feels to me that you want a mix of vscode and IntelliJ keybindings. In that case, you can further refine the keybindings defined by the extension. If you really want to change things, you can. User keybindings get the highest priority over extension keybindings and you can configure everything to your liking. Just take a look at our [extensive documentation](https://code.visualstudio.com/docs/getstarted/keybindings).

Here is the current keybinding defined for `ctrl+=` by the extension: https://github.com/kasecato/vscode-intellij-idea-keybindings/blob/c55732c6d3fd4649bebf4d3cea61220f34665eba/package.json#L215-L221
```
            {
                "key": "ctrl+=",
                "mac": "cmd+=",
                "command": "editor.unfold",
                "when": "editorFocus",
                "intellij": "Expand code block"
            },
```

Basically, `ctrl+=` unfolds when focus is in an editor and zooms in when focus is anywhere else. That is why the lop level menus render the keybinding next to the actions, because there is a possibility to trigger the built-in keybinding. 

I'm not exactly sure what the complaint is here. If you like the IntelliJ IDEA keybindings but dislike this particular keybinding, you can very easily remove them. e.g. you can edit your `keybindings.json` file and remove them:
(You can also remove them with our keybindings UI editor ctrl+k ctrl+s by right clicking on them)

```
{ "key": "ctrl+=", "command": "-editor.unfold" }
```

**Comment by Chealer** [comment]
Thank you @alexdima , 
> I have considered the first 3 suggestions, but we cannot implement them. I think there is a misunderstanding of expectations here. 

I agree, great point. I shall clarify this.

> When you install the IntelliJ IDEA keymap, you basically agree to get the keybindings as close as possible to IntelliJ IDEA. If IntelliJ IDEA does not zoom in on `ctrl+=`, then vscode configured with the IntelliJ IDEA keymap will not zoom in on `ctrl+=`. There is no reason to prevent or warn because the feature works as expected.

I understand how you may perceive it this way, but I did not actually agree to that. IntelliJ IDEA and VScode are different applications offering different features. I do not expect an IntelliJ IDEA shortcut to have an equivalent effect in VScode if VScode has no equivalent feature. I also do not expect a certain feature to have - for example - no shortcut just because IntelliJ IDEA does not offer a shortcut for it.

Expectations could vary depending on who you ask, but basically, if I install that extension, all I expect is my migration (or co-usage) to be eased (at least in terms of keybindings). If an IntelliJ shortcut is unassigned in VScode, I do expect the extension to assign that shortcut if VScode has a corresponding function. But I do not **expect** that an existing VScode shortcut would see its effect changed. I would "expect" such conflicts to be examined one by one and the best default weighed case by case. All I basically expect from the extension is to make VScode more predictable to me. I do not expect VScode's behavior to become misleading or any less predictable.
 
> I'm not exactly sure what the complaint is here.

I am sorry about that, my understanding of the problem has evolved and I recognize that has not been too clear. Thank you for the remark, and let me try to clarify.

My understanding is that VScode has a general extensibility problem with regards to key bindings. It allows extensions to change the behavior of shortcuts and sometimes fails to react properly when an extension does that. Solving this may require a change to the extension API, or simply one of the changes I listed in [my previous comment](https://github.com/microsoft/vscode/issues/110254#issuecomment-729756195).

However, I raised this issue in the specific context of the IntelliJ IDEA Keybindings extension's zoom shortcuts, and did not notice the wider issue anywhere else so far. I will be satisfied with a fix focused on the narrower issue if someone thinks this suffices or feels more comfortable implementing that than a systemic fix.

**Comment by alexdima** [comment]
The entire purpose of a keymap extension is to change existing keybindings. Here are some of the keybindings that are changed by the extension:
* [`ctrl+p`](https://github.com/kasecato/vscode-intellij-idea-keybindings/blob/c55732c6d3fd4649bebf4d3cea61220f34665eba/package.json#L57-L60) which will trigger parameter hints instead of opening the Command Palette
* [`ctrl+q`](https://github.com/kasecato/vscode-intellij-idea-keybindings/blob/c55732c6d3fd4649bebf4d3cea61220f34665eba/package.json#L64-L67) which will show the hover instead of opening the quick pick
* [`ctrl+w`](https://github.com/kasecato/vscode-intellij-idea-keybindings/blob/c55732c6d3fd4649bebf4d3cea61220f34665eba/package.json#L111-L114) which will expand the selection instead of closing the active editor
* [`ctrl+d`](https://github.com/kasecato/vscode-intellij-idea-keybindings/blob/c55732c6d3fd4649bebf4d3cea61220f34665eba/package.json#L160-L163) which will copy lines down instead of adding a selection to the next find match
* ... and on and on and on including `ctrl+=`

All of these are examples of built-in keybindings being overwriten by the extension.

I have really tried all my best to extract something actionable from this discussion, but I have to say I don't have any more energy to debate this. This is literally how keymap extensions work, this is the purpose of the `keybindings` extension point, the extension appears to be doing everything correctly, and IMHO there is nothing wrong here and everything works exactly as designed.

**Comment by Chealer** [comment]
> The entire purpose of a keymap extension is to change existing keybindings. 

Unless one defines "existing keybindings" very largely (meaning "all keybindings"), that is incorrect. The purpose of a keymap extension is to provide keybindings offered by the imitated software. In a very lucky case, adding keybindings would suffice to do that. In a very unlucky case, this would involve nothing but changes to (existing) keybindings. In reality, each extension falls somewhere different on a spectrum between these extremes.

> [...]
> 
> I have really tried all my best to extract something actionable from this discussion, but I have to say I don't have any more energy to debate this. This is literally how keymap extensions work, this is the purpose of the `keybindings` extension point, the extension appears to be doing everything correctly, and IMHO there is nothing wrong here and everything works exactly as designed.

Yes, I see there is an understanding issue, but [the above comment](https://github.com/microsoft/vscode/issues/110254#issuecomment-729756195) already lists several ways this can be fixed, and I doubt I can help much more at this point. If there is any information missing, please specify.
